topostats/tracing/disordered_tracing.py

```diff
@@ -0,0 +1,708 @@
+
+        LOGGER.debug(f"[{self.filename}] Performing Disordered Tracing")
+
+    def trace_dna(self):
+        """Perform the DNA skeletonisation and cleaning pipeline."""
+        # LOGGER.info(f"[{self.filename}] : mask_smooth_params : {self.mask_smoothing_params=}")
+        self.smoothed_mask = self.smooth_mask(self.mask, **self.mask_smoothing_params)
+        self.skeleton = getSkeleton(
+            self.smoothed_mask,
+            self.mask,
+            method=self.skeletonisation_params["method"],
+            height_bias=self.skeletonisation_params["height_bias"],
+        ).get_skeleton()
+        self.pruned_skeleton = prune_skeleton(self.smoothed_mask, self.skeleton, **self.pruning_params.copy())
+        self.pruned_skeleton = self.remove_touching_edge(self.pruned_skeleton)
+        self.disordered_trace = np.argwhere(self.pruned_skeleton == 1)
+
+        if self.disordered_trace is None:
+            LOGGER.info(f"[{self.filename}] : Grain failed to Skeletonise")
+        elif len(self.disordered_trace) < self.min_skeleton_size:
+            self.disordered_trace = None
+
+    def re_add_holes(
+        self, orig_mask: npt.NDArray, new_mask: npt.NDArray, holearea_min_max: list = (4, np.inf)
+    ) -> npt.NDArray:
+        """
+        Restore holes in masks that were occluded by dilation.
+
+        As Gaussian dilation smoothing methods can close holes in the original mask, this function obtains those holes
+        (based on the general background being the first due to padding) and adds them back into the smoothed mask. When
+        paired with smooth_mask, this essentially just smooths the outer edge of the mask.
+
+        Parameters
+        ----------
+        orig_mask : npt.NDArray
+            Original mask.
+        new_mask : npt.NDArray
+            New mask.
+        holearea_min_max : list
+            List of minimum and maximum hole area (in pixels).
+
+        Returns
+        -------
+        npt.NDArray
+            Smoothed mask with holes restored.
+        """
+        holesize_min_px = holearea_min_max[0] / ((self.pixel_to_nm_scaling / 1e-9) ** 2)
+        holesize_max_px = holearea_min_max[1] / ((self.pixel_to_nm_scaling / 1e-9) ** 2)
+        holes = 1 - orig_mask
+        holes = label(holes)
+        sizes = [holes[holes == i].size for i in range(1, holes.max() + 1)]
+        holes[holes == 1] = 0  # set background to 0
+
+        for i, size in enumerate(sizes):
+            if size < holesize_min_px or size > holesize_max_px:  # small holes may be fake are left out
+                holes[holes == i + 1] = 0
+        holes[holes != 0] = 1
+
+        # compare num holes in each mask
+        holey_smooth = new_mask.copy()
+        holey_smooth[holes == 1] = 0
+
+        return holey_smooth
+
+    @staticmethod
+    def remove_touching_edge(skeleton: npt.NDArray) -> npt.NDArray:
+        """
+        Remove any skeleton points touching the border (to prevent errors later).
```

Comment:
What sort of errors does this throw?

It might be useful to use these as a basis for tests to ensure that this function does indeed remove them.

Reply:
No idea as this is legacy tracing code and the error doesn't occur in this new module so we might have to test when the pipline is complete. My intuition tells me an indexError when trying to order a linear segment as the `get_node_count_and_area()` func attempts to find the 3x3 area around the node which would not be possible.

In addition, the above would only happend when the `pad_width` is 0, otherwise this function is kind of redundant (unless looking at large webbed structures but you'd usually want to keep these).